### PR TITLE
Rmd refactor (part 2)

### PR DESCRIPTION
**Pull request by pdiakumis**
Refactoring mainly the input data reading and list setup:

- `DESCRIPTION`: add a couple deps, set up testthat
- `arriba.R`: refactored this with functions for breaking up the pdf into png files, reading the TSV, writing the fusion names for piedb, handling NULL, added tests.
- `R/clinical_info.R`: handling the xlsx reading here, added test
- `R/dragen_fusions.R`: handling the `fusion_candidates.final` reading here, handling NULL, added tests.
- `R/prepare2write.R`: might get rid of this eventually.
- `R/refdata.R`: convenience functions for setting up the refdata and genes list, added tests.
- `R/salmon.R`: handling salmon `quant.sf` reading via `tximport`, handling NULL, added (silly) test.
- `R/sample_data.R`: wrapper for handling arriba, salmon, dragen fusions, clinical info reading, added test. Output is a list with those input files parsed (or NULL if not provided).
- `R/utils.R`: added `mkdir` func (and included `%!in%` in here)
- `inst/cli/rnasum.R`: just reordered the options alphabetically, and think I removed one or two
- `man/`: just the automatically generated help files
- `tests/`: contains the tests. Tests are written above the `R/` funcs under the `@testexamples`. This is via the awesome `{roxytest}` functionality (https://github.com/mikldk/roxytest), which makes it super convenient to integrate short tests with your code. Need to set that up locally to make the docs work, just ping me when you want to try it out.
- `vignettes/development.Rmd`: just moved what was in the report itself to this.

For the report Rmd itself, mostly refactoring changes. The report won't work as-is right now, need to adjust the variable names throughout, but I don't want to drag this PR out any longer or else it will get out of hand.

### CODE REVIEW

**Comment by pdiakumis** [comment]
Okay, parking this for now, added some more changes:

- Using `{gpgr}`'s PURPLE and Manta parsers (the old code was probably from 2019).
- Added `pcgr_tiers_tsv_read` for reading the PCGR tiers TSV.
- Added `REFERENCE_DATASETS` data list that holds the ref dataset info
- Refactored CLI script to simplify bits and avoid duplicating info
- Added parsers for `FusionGDB`, `cancer_biomarkers` etc.
- Refactored `R/combineDatasets.R` a little bit

File: R/combineDatasets.R [hunk 1]
```diff
@@ -1,93 +1,75 @@
...
-  target.sample <- data.frame(sample_name, sample_name)
-  names(target.sample) <- names(target.comb)
-  rownames(target.sample) <- sample_name
-  target.comb <- rbind( target.comb, target.sample )
-
-  ##### Make syntactically valid names
-  rownames(target.comb) <- make.names(rownames(target.comb))
+combineDatasets <- function(sample_name, sample_counts, ref_data, report_dir) {
+  assertthat::assert_that(!is.null(sample_counts), msg = "The 'sample_counts' cannot be NULL!")
+  .read_target <- function(targ, ds) {
```
Comment (skanwal) [inline R/combineDatasets.R:16]: For my information, any particular reason to use `.name`?

File: R/dragen_fusions.R [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#' (a <- dragen_fusions_read(x))
+#' @testexamples
+#' expect_equal(colnames(a)[ncol(a)], "ReadNames")
+#' expect_equal(colnames(a)[1], "gene1")
+#' expect_null(dragen_fusions_read(NULL))
+#' @export
+dragen_fusions_read <- function(x = NULL) {
+  if (is.null(x)) {
+    return(NULL)
+  }
```
Comment (skanwal) [inline R/dragen_fusions.R:20]: Is this the test you had mentioned in the PR description?

File: R/sample_data.R [hunk 3]
```diff
@@ -0,0 +1,127 @@
+#' Read Sample Data
+#'
+#' Reads sample data, including Arriba fusions, Arriba plots, Salmon counts,
+#' DRAGEN fusions, and clinical info.
```
Comment (skanwal) [inline R/sample_data.R:4]: For now, the code is sticking with tool names in variables?

Comment (skanwal) [reply R/sample_data.R:4]: Checking as I recall you mentioned this was something you were updating.

File: tests/testthat/test-roxytest-testexamples-arriba.R [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Generated by roxytest: Do not edit by hand!
+
+# File R/arriba.R: @testexamples
+
+test_that("Function arriba_tsv_read() @ L14", {
+  
+  x <- system.file("rawdata/test_data/dragen/arriba/fusions.tsv", package = "RNAsum")
```
Comment (skanwal) [inline tests/testthat/test-roxytest-testexamples-arriba.R:7]: oh I see these are the tests, probably?

File: vignettes/development.Rmd [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+title: "Development"
+output: rmarkdown::html_vignette
+date: "`r Sys.Date()`"
+vignette: >
+  %\VignetteIndexEntry{Development}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+Notes on the RNAsum Rmd structure for developers
```
Comment (skanwal) [inline vignettes/development.Rmd:11]: Nice! Really handy.

**Review by skanwal — COMMENTED** [review]
These were a lot of changes to go through. Firstly, thank you! - secondly, It'll help me follow through the updates better if we can make these PRs smaller in the future please. 

Quick question, where is the functionality of `sv_prioritize` moved to? I probably missed it while skimming through the code.

Also, I noted few params have been removed/added from rmd, which are still using input tool names. What is the major motivation for that?

**Comment by pdiakumis** [comment]
Hey, totally agree that there are a lot of changes in this PR, sorry! I think it's a bullet we need to bite though. That's the main reason this refactor is absolutely necessary imho: in order to make a small change in e.g. the fusion input parsing, you need to change things in multiple parts throughout the report (e.g. imagine if you want to simply change the name of a column).
Having the main setup refactored into a set of functions helps decrease that barrier. And that's what this PR is focused on: set up functions to read fusions/SVs/SNVs/CNVs/gene panels etc., set up functions to read in the refdata, manage those in a separate location, add tests for all. The changes here are related to the first ~500 lines of code in the Rmd, and the CLI script. I did try hitting the break last week but there was just too much related stuff (obviously I was in the zone, and you always tend to drag it on when you're in the zone!). The test data also needs to be updated at some stage (pcgr/manta/purple files are all old, some reference files are windows-generated etc.).

- You can basically ignore all the files under `tests/` and `man/` (there's 25 changed files right there), and just focus on the `R/` stuff. The CLI script and Rmd are a work in progress. The CLI script was simplified to use the `opt` list directly, instead of duplicating variable assignment in the rmarkdown params list.
  - The `tests/` folder contains that stuff you see above the functions in `R/` under the `@testexamples` sections. The way it works is, you write your tests above the function definition itself, then `{roxytest}` takes those and automatically adds them to `tests/`. It saves a ton of time and helps you actually write tests.

- I've included tool-specific params for flexibility. If someone external wants to use this, they probably won't have the same file structure we get from umccrise/DRAGEN WTS, so they can provide the required inputs individually (or not at all). The `dragen_rnaseq` param has been removed for now since it was only used to point to hard-coded file paths - I'll re-add that if needed and treat it as optional, the same way I'm handling the `umccrise` param. In short, you first look if those tool-specific params are specified, and if they're not, then look for the patterns under `umccrise`, and if they're not there either just return NULL.

- The `sv_prioritize` functionality has been replaced with `gpgr::process_sv` - look for the `manta_tsv` section in `R/sample_data.R` (`read_wgs_data` func).

Ultimately, we're gonna run a set of tests at the end to make sure the output matches with the master version, so I wouldn't focus too much on specific changes.

File: R/combineDatasets.R (hunk 1 above)
Comment (pdiakumis) [reply R/combineDatasets.R:16]: Not really, I guess it makes it clear this is an internal func and it also makes it hidden in the RStudio env pane when testing (which means less clutter).

File: vignettes/development.Rmd (hunk 5 above)
Comment (pdiakumis) [reply vignettes/development.Rmd:11]: Those were copied from the Rmd itself, not taking any credit (will also need to be updated!)

File: R/dragen_fusions.R (hunk 2 above)
Comment (pdiakumis) [reply R/dragen_fusions.R:20]: These NULL checks essentially handle the case when a param is not provided.
Here's an example:

```r
param <- list(
  a = "one",
  b = "two"
)

param$a # "one"
param$dragen_fusion # NULL
```

So then you have `df <- dragen_fusions_read(params$dragen_fusion)` returning NULL, and then you have your chunk runner being:

```
runDragenFusionChunk <- !is.null(df) # FALSE
```

That's the general idea ;-)

File: vignettes/development.Rmd (hunk 5 above)
Comment (skanwal) [reply vignettes/development.Rmd:11]: The comment was on the "Notes on the RNAsum Rmd...." bit not the content, as I found the idea to have it all in `development.Rmd` useful.

**Comment by skanwal** [comment]
Definitely, not questioning the value of these changes, which are needed at this point. 
Trying to think how to best make these changes manageable.

> The test data also needs to be updated at some stage
Yep, that has been on my mind, ideally with the latest dragen results.

`dragen_rnaseq` param is used to point to the RNAseq results from Dragen. This is the one we are using currently, so will definitely need that. Again, this will be updated to something more general (without dragen/bcbio prefix), so that it just points to analysis outputs from rnaseq analysis.

**Comment by pdiakumis** [comment]
> Trying to think how to best make these changes manageable.

I think they'll become more manageable from now on. The tricky part was getting the bulk of the data parsing refactored. Now can focus on other, more targeted aspects of the report.

**Review by skanwal — APPROVED** [review]

**Comment by skanwal** [comment]
All good and approved!
Thanks again.

**Pull request merged**